### PR TITLE
DOP-2732: generate search manifests from AST

### PR DESCRIPTION
**Pull request by schmalliso**
Submitting for an initial review, with the understanding that there's some cleaning up to do in my enthusiastic print statements. I'm also uncertain as to whether we'd want to keep the `source` data from which I've generated most of the BSON in here, or if we'd just want to leave the BSON there and remove it. I'm fine either way.

Obviously, there's also some code to delete (the other manifest-generation stuff), but I wanted to leave that be _for now_.

Generated manifest is here: [test-manifest.json.zip](https://github.com/mongodb/mut/files/9001622/test-manifest.json.zip) because GitHub won't let me attach a JSON doc directly, I zipped it. HACKY!

### CODE REVIEW

File: mut/index/test_SnootyManifest.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
+from base64 import decode
+from pydoc import Doc, doc
+from bson import decode_all
+from json import loads
+from pathlib import Path
+
+from mut.index.SnootyManifest import Document, generate_manifest
+
+
+ROOT_PATH = Path("/Users/allison/snooty/mut/mut/test_data_index/documents")
```
Comment (schmalliso) [inline mut/index/test_SnootyManifest.py:10]: Clearly I'll need to fix this.

File: mut/index/SnootyManifest.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+from typing import Any, Dict, Optional, List
+
+class Document:
+    '''Return indexing data from a page's AST for search purposes.'''
+    def __init__(self, data) -> None:
+
+        self.tree = data[0]
+
+        self.robots, self.keywords, self.description = self.findMetadata()
```
Comment (i80and) [inline mut/index/SnootyManifest.py:18]: Python method names should be `snake_case`

File: mut/index/SnootyManifest.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    def export(self) -> str:
+        '''Return the manifest as json.'''
+        manifest = {
+            "url": self.url,
+            "includeInGlobalSearch": self.globally,
+            "documents": self.documents
+        }
+        return dumps(manifest, indent=4)
+
+def process_snooty_manifest_bson(path: str) -> Any:
```
Comment (i80and) [inline mut/index/SnootyManifest.py:176]: I'm increasingly a fan of using pathlib.Path whenever possible.

Bonus: then you can just do `data = decode_all(path.read_bytes())`

File: mut/index/SnootyManifest.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+            manifest.add_document(bson_doc)
+        return manifest
+
+def get_ast_list(walk_dir: str) -> List[str]:
+    '''Get full list of BSON paths that need to be processed,
+       get rid of files that don't need to be indexed, but exist as AST.'''
+    ast_source_paths: List[str] = []
+
+    for root, dirs, files in walk(walk_dir):
+        if 'images' in dirs:
```
Comment (i80and) [inline mut/index/SnootyManifest.py:200]: I'd prefer to make this less branch-y. e.g.
```
for forbidden_name in ["images", "includes", "sharedinclude", ".DS_Store"]:
    if forbidden_name in dirs:
        dirs.remove(forbidden_name)
```

If you want to get FANCY, you could even do a set:
```
for forbidden_name in set(["images", "includes", "sharedinclude", ".DS_Store").intersection(dirs):
    dirs.remove(forbidden_name)
```

File: mut/index/SnootyManifest.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        jsonpath_expr = parse('$..children[?(@.type==\'code\')]')
+        results = jsonpath_expr.find(self.tree)
+        code_contents = []
+        for r in results:
+            lang = r.value.get("lang", None)
+            code_contents.append({"lang": lang, "value": r.value["value"]})
+        
+        return code_contents
+
+    def findHeadings(self):
```
Comment (i80and) [inline mut/index/SnootyManifest.py:52]: the return type of this appears to be `Tuple[Optional[str], Optional[List[str]]`.

File: mut/index/SnootyManifest.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        title = headings[0]
+        headings.pop(0)
+        return title, headings
+
+    def deriveSlug(self):
+        logger("\t Deriving slug")
+        page_id = self.tree["page_id"]
+        return page_id
+
+    def derivePreview(self):
```
Comment (i80and) [inline mut/index/SnootyManifest.py:77]: is this returning `Optional[str]`?

File: mut/index/SnootyManifest.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+        # If page has no title, do not index.
+        if self.title == None:
+            noindex = True
+            reasons.append("This page has 0 headings, not even the H1")
+
+        logger("\tnoindex: {}".format(noindex))
+        return noindex, reasons
+
+    def export(self) -> Optional[Dict[str, Any]]:
```
[the remark this answers is not in the document]
Comment (schmalliso) [reply mut/index/SnootyManifest.py:136]: fancy!

File: mut/index/SnootyManifest.py (hunk 4 above)
Comment (schmalliso) [reply mut/index/SnootyManifest.py:200]: LOVE IT.

File: mut/index/SnootyManifest.py (hunk 3 above)
Comment (schmalliso) [reply mut/index/SnootyManifest.py:176]: I think I used that in the tests, but somehow didn't port it over here.

File: mut/index/SnootyManifest.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+            dirs.remove('includes')
+        if 'sharedinclude' in dirs:
+            dirs.remove('sharedinclude')
+        if '.DS_Store' in files:
+            files.remove('.DS_Store')
+        for filename in files:
+            ast_source_paths.append(join(root, filename))
+    return ast_source_paths
+
+def logger(message: str) -> None:
```
[the remark this answers is not in the document]
Comment (schmalliso) [reply mut/index/SnootyManifest.py:212]: lol. yes. I was using this dodgy little bugger just so it was easy to turn on and off when testing, but have updated to actual logger :D

File: mut/index/SnootyManifest.py (hunk 4 above)
Comment (schmalliso) [reply mut/index/SnootyManifest.py:200]: SO MUCH CLEANER

**Comment by schmalliso** [comment]
@i80and Updated (finally). Note that you can view the updates from the feedback patch using the 'Changes from all commits' dropdown on the top left of the Files Changed page so you don't have to wade through all the linting.

File: mut/index/test_SnootyManifest.py (hunk 1 above)
Comment (i80and) [reply mut/index/test_SnootyManifest.py:7]: Question mark :D

**Review by i80and — CHANGES_REQUESTED** [review]

File: mut/index/test_SnootyManifest.py (hunk 1 above)
Comment (schmalliso) [reply mut/index/test_SnootyManifest.py:7]: So realtalk, I'm not really sure *how* to. Will ping you 😁

File: mut/index/test_SnootyManifest.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
+from bson import decode_all
+from json import loads
+from pathlib import Path
+from mut.index.SnootyManifest import ManifestEntry, Document, generate_manifest
+
+
+ROOT_PATH = Path("../test_data_index/documents")
```
Comment (schmalliso) [inline mut/index/test_SnootyManifest.py:7]: @i80and I used my brain, this should work non?

File: mut/index/SnootyManifest.py [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        self.noindex, self.reasons = self.get_noindex()
+
+    def find_paragraphs(self) -> str:
+        logger.debug("Finding paragraphs")
+        # NB: paragraphs include "paragraph" nodes within tables
+        jsonpath_expr = parse("$..children[?(@.type=='paragraph')]..value")
+        results = jsonpath_expr.find(self.tree)
+
+        # Appending to then joining an array is faster than repeatedly concatenating strings
+        str_list = []
```
Comment (i80and) [inline mut/index/SnootyManifest.py:49]: you could idiomatically do this as a list comprehension:
```
str_list = [r.value for r in results]
```

**Review by i80and — APPROVED** [review]
One educational note, but if you've tested this, lgtm!

File: mut/index/SnootyManifest.py (hunk 10 above)
Comment (schmalliso) [reply mut/index/SnootyManifest.py:49]: that is slick, but I am also ready to close this out :D thanks for the learn <3

**Pull request merged**